launcher/envvars.go

```diff
@@ -0,0 +1,206 @@
+	log "github.com/sirupsen/logrus"
+)
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+func (launcher *Launcher) setEnvVars(
+	envVarsInfo []cloudprotocol.EnvVarsInstanceInfo,
+) (envVarsStatus []cloudprotocol.EnvVarsInstanceStatus) {
+	envVarsStatus = make([]cloudprotocol.EnvVarsInstanceStatus, 0, len(envVarsInfo))
+
+	now := time.Now()
+
+	for _, envVarInfo := range envVarsInfo {
+		envVarStatus := cloudprotocol.EnvVarsInstanceStatus{
+			InstanceFilter: envVarInfo.InstanceFilter,
+		}
+
+		for _, envVar := range envVarInfo.EnvVars {
+			log.WithFields(
+				instanceFilterLogFields(envVarInfo.InstanceFilter, log.Fields{
+					"var": envVar.Variable,
+					"ttl": envVar.TTL,
+					"id":  envVar.ID,
+				}),
+			).Debug("Override env var")
+
+			if envVar.TTL != nil && envVar.TTL.Before(now) {
+				err := aoserrors.New("environment variable expired")
+
+				envVarStatus.Statuses = append(envVarStatus.Statuses,
+					cloudprotocol.EnvVarStatus{ID: envVar.ID, Error: err.Error()})
+
+				log.WithField("id", envVar.ID).Errorf("Error overriding environment variable: %s", err)
+
+				continue
+			}
+
+			envVarStatus.Statuses = append(envVarStatus.Statuses, cloudprotocol.EnvVarStatus{ID: envVar.ID})
+		}
+
+		envVarsStatus = append(envVarsStatus, envVarStatus)
+	}
+
+	launcher.currentEnvVars = envVarsInfo
+
+	if err := launcher.storage.SetOverrideEnvVars(envVarsInfo); err != nil {
+		return setEnvVarsErr(envVarsStatus, err)
+	}
+
+	return envVarsStatus
+}
+
+func setEnvVarsErr(
+	envVarsStatus []cloudprotocol.EnvVarsInstanceStatus, err error,
+) []cloudprotocol.EnvVarsInstanceStatus {
+	var errStr string
+
+	if err != nil {
+		errStr = err.Error()
+	}
+
+	for _, envVarStatus := range envVarsStatus {
+		for i, status := range envVarStatus.Statuses {
+			if status.Error == "" {
+				envVarStatus.Statuses[i] = cloudprotocol.EnvVarStatus{
+					ID: status.ID, Error: errStr,
+				}
+			}
+		}
+	}
+
+	return envVarsStatus
+}
+
+func (launcher *Launcher) getInstanceEnvVars(instance InstanceInfo) (envVars []string) {
+	for _, envVarInfo := range launcher.currentEnvVars {
+		if (envVarInfo.ServiceID == instance.ServiceID) &&
+			(envVarInfo.SubjectID == nil || *envVarInfo.SubjectID == instance.SubjectID) &&
+			(envVarInfo.Instance == nil || *envVarInfo.Instance == instance.Instance) {
+			for _, envVar := range envVarInfo.EnvVars {
+				envVars = append(envVars, envVar.Variable)
+			}
+		}
+	}
+
+	return envVars
+}
+
+func (launcher *Launcher) removeOutdatedEnvVars() {
+	var (
+		now            = time.Now()
+		updatedEnvVars = make([]cloudprotocol.EnvVarsInstanceInfo, 0, len(launcher.currentEnvVars))
+		updated        = false
+	)
+
+	for _, item := range launcher.currentEnvVars {
+		updatedItems := make([]cloudprotocol.EnvVarInfo, 0, len(item.EnvVars))
```

Comment:
To summarize my point of append vs array[i] = val on preallocated slice: append doesn't allocate additional memory in this case, just perform additional check inside. But it updates the len properly in case you have some conditions in the loop. So, for me it is more safe to use append than array[i] = val except you have very tough time constraints.